ui_framework/src/components/code_editor/code_editor.js

```diff
@@ -0,0 +1,98 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import classNames from 'classnames';
+import AceEditor from 'react-ace';
+
+import { htmlIdGenerator, keyCodes } from '../../../services';
+
+export class KuiCodeEditor extends Component {
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      isHintActive: true
+    };
+    this.idGenerator = htmlIdGenerator();
+  }
+
+  aceEditorRef = (aceEditor) => {
+    if (aceEditor) {
+      this.aceEditor = aceEditor;
+      aceEditor.editor.textInput.getElement().tabIndex = -1;
+      aceEditor.editor.textInput.getElement().addEventListener('keydown', this.onKeydownAce);
+    }
+  };
+
+  stopEditing() {
+    this.setState({ isHintActive: true });
+  }
+
+  onKeydownAce = (ev) => {
+    if (ev.keyCode === keyCodes.ESCAPE) {
+      ev.preventDefault();
+      ev.stopPropagation();
+      this.stopEditing();
+      this.refs.kuiCodeEditorHint.focus();
+    }
+  }
+
+  onBlurAce = (...args) => {
+    this.stopEditing();
+    if (this.props.onBlur) {
+      this.props.onBlur(...args);
+    }
+  };
+
+  onKeyDownHint = (ev) => {
+    if (ev.keyCode === keyCodes.ENTER) {
+      ev.preventDefault();
+      this.startEditing();
+    }
+  };
+
+  startEditing = () => {
+    this.setState({ isHintActive: false });
+    this.aceEditor.editor.textInput.focus();
+  }
+
+  render() {
+    const { width, height } = this.props;
+    const classes = classNames('kuiCodeEditorKeyboardHint', {
+      'kuiCodeEditorKeyboardHint-isInactive': !this.state.isHintActive
+    });
+    return (
+      <div
+        className="kuiCodeEditorWrapper"
+        style={{ width, height }}
+      >
+        <div
+          className={classes}
+          id={this.idGenerator('codeEditor')}
+          ref="kuiCodeEditorHint"
```

Comment:
Don't use string refs, use callback refs (see https://facebook.github.io/react/docs/refs-and-the-dom.html)